### PR TITLE
Add support for file uploads

### LINKED ISSUE #146: Feature Proposal: File Upload
Hej guys,

I'd love to use artillery for testing file uploads as well. I *do* know there's an open issue (https://github.com/shoreditch-ops/artillery/issues/106).

But as this is a thing that needs to be implemented here, plus I couldn't find an already discussed proposal.
I'd love to get some feedback, as I want start working on this 🙃 

<hr>

# File Upload

## Rationale

Uploading files is a frequent HTTP use-case, but currently not supported very well within `artillery-core`.

To enable user to follow a unified approach, this proposal adds a cURL-style way of indicating file uploads.

In the background each file be passed to `request` using a `ReadableStream` to allow tests with potentially huge file sizes.

## Changes

Indicate file payload by prepending the path with `@`, e.g. `@path/to/file.md`.

**Input types:**

* single file
* list of files
* list of objects with `key` and `path` values (form multipart)

**File resolution:**

If the `@` sign (file indicator) is immediately followed by a `/`, file resolution happens from root of disk.

Otherwise files will be tried to be resolved in following order:

1. `process.cwd` + your path
2. `__dirname` + your path

**Backwards compatibility: :warning:**

People who previously started their custom body payload with an at-sign (`@`) will need to adjust their descriptors.

I can't really figure, why you would do that - but as it wasn't disallowed previously I'd rather go with a semver valid major bump / breaking change.

## Examples of Input

**Single File**

```yaml
post:
  url: "/test"
  body: "@selfie.jpg"
```

**Multiple Files**

```yaml
post:
  url: "/test"
  body:
    - "@selfie.jpg"
    - "@selfie2.png"
```

**Multiple Files - Multipart Form**

```yaml
post:
  url: "/test"
  body:
    - key: foo
      path: "@selfie.jpg"
    - key: bar
      path: "@selfie2.png"
```

### PR DESCRIPTION
**Pull request by zcei**
## Examples of Input

Include them in config:

```yaml
config:
  files:
    - "@./path/to/file"
    - "@../path/to/other/file"
```

Reference them from within your scenario

**Single File**

```yaml
post:
  url: "/test"
  body: "@./path/to/file"
```

**Multiple Files - Multipart Form**

```yaml
post:
  url: "/test"
  formData:
    foo: "@./path/to/file"
    bar: "@../path/to/other/file"
```

Closes #146

### CODE REVIEW

**Comment by hassy** [comment]
Cool, cheers. So this needs a corresponding change in the CLI to read those files right?

I'm wondering if we could get rid of `config.files` altogether, as it would seem odd from a user's point of view to need to declare the files which will be used and then reference the same path again.

**Comment by zcei** [comment]
Yes, CLI needs to be changes, PR for this is incoming (hopefully this week)

If you're fine with traversing the whole `scenarios` object we could filter out the file paths within CLI and pass the ReadStream instead. (assuming you agree that `artillery-core` shouldn't do the I/O)

**Comment by hassy** [comment]
@zcei I'll merge this PR - we can tweak how it works if needed later. Cheers

**Comment by armand1m** [comment]
It would be nice to have this on docs! Can I propose a pull request with the docs updated?

**Comment by zcei** [comment]
I'd love to see it documented 
Better to file the PR in [artillery-docs](https://github.com/shoreditch-ops/artillery-docs), though

You can use this PR for the CLI as a reference: https://github.com/shoreditch-ops/artillery/pull/265

**Comment by armand1m** [comment]
@zcei Thanks, I'll write it down as soon as possible and submit the PR to the artillery-docs project.

File: lib/engine_http.js [hunk 1]
```diff
@@ -172,6 +174,20 @@ HttpEngine.prototype.step = function step(requestSpec, ee, opts) {
         {});
     }
 
+    if (params.formData) {
+      requestParams.formData = _.reduce(
+        requestParams.formData,
+        function (acc, v, k) {
+          const reference = fileReference(v, config);
```
Comment (cskru) [inline lib/engine_http.js]: Aren't you passing formData with read stream?
Like this:
            formData: {
                foo: {
                    value: fs.createReadStream(filePath),
                    options: {
                        filename: fileName,
                        contentType: 'application/octet-stream'
                    }
                }
            }

**Pull request merged**